File: config/src/main/java/com/typesafe/config/impl/ConfigLong.java
/**
 *   Copyright (C) 2011-2012 Typesafe Inc. <http://typesafe.com>
 */
package com.twitter_typesafe.config.impl;

import java.io.ObjectStreamException;
import java.io.Serializable;

import com.twitter_typesafe.config.ConfigOrigin;
import com.twitter_typesafe.config.ConfigValueType;

final class ConfigLong extends ConfigNumber implements Serializable {

    private static final long serialVersionUID = 2L;

    final private long value;

    ConfigLong(ConfigOrigin origin, long value, String originalText) {
        super(origin, originalText);
        this.value = value;
    }

    @Override
    public ConfigValueType valueType() {
        return ConfigValueType.NUMBER;
    }

    @Override
    public Long unwrapped() {
        return value;
    }

    @Override
    String transformToString() {
        String s = super.transformToString();
        if (s == null)
            return Long.toString(value);
        else
            return s;
    }

    @Override
    protected long longValue() {
        return value;
    }

    @Override
    protected double doubleValue() {
        return value;
    }

    @Override
    protected ConfigLong newCopy(ConfigOrigin origin) {
        return new ConfigLong(origin, value, originalText);
    }

    // serialization all goes through SerializedConfigValue
    private Object writeReplace() throws ObjectStreamException {
        return new SerializedConfigValue(this);
    }
}

File: config/src/main/scala/com/typesafe/config/impl/ConfigLong.scala
/**
 *   Copyright (C) 2011-2012 Typesafe Inc. <http://typesafe.com>
 */
package com.typesafe.config.impl

import java.{ lang => jl }
import java.io.ObjectStreamException
import java.io.Serializable
import com.typesafe.config.ConfigOrigin
import com.typesafe.config.ConfigValueType

@SerialVersionUID(2L)
final class ConfigLong(
    origin: ConfigOrigin,
    val value: Long,
    originalText: String)
    extends ConfigNumber(origin, originalText)
    with Serializable {

    override def valueType: ConfigValueType = ConfigValueType.NUMBER

    override def unwrapped: jl.Long = value

    override def transformToString: String = {
        val s = super.transformToString
        if (s == null) jl.Long.toString(value) else s
    }
    override def longValue: Long = value

    override def doubleValue: Double = value

    override def newCopy(origin: ConfigOrigin) =
        new ConfigLong(origin, value, originalText)

    // serialization all goes through SerializedConfigValue
    @throws[ObjectStreamException]
    private def writeReplace(): jl.Object = new SerializedConfigValue(this)
}
